### PR TITLE
Add hcloud builder

### PR DESCRIPTION
**Pull request by LKaemmerling**
Hello guys,

we at Hetzner Cloud want to provide an official hcloud builder for packer, the same way we provide an official hcloud provider for terraform.

### CODE REVIEW

File: website/source/docs/builders/hetzner-cloud.html.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+---
+description: |
+    The Hetzner Cloud Packer builder is able to create new images for use with
+    DigitalOcean. The builder takes a source image, runs any provisioning
```
Comment (rickard-von-essen) [inline website/source/docs/builders/hetzner-cloud.html.md]: DI

File: website/source/docs/builders/hetzner-cloud.html.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+[communicator](/docs/templates/communicator.html) can be configured for this
+builder.
+
+### Required:
+
+-   `token` (string) - The client TOKEN to use to access your account. It
+    can also be specified via environment variable `HCLOUD_TOKEN`,
+    if set.
+
+-   `image` (string) - 	ID or name of location to create server in.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/hetzner-cloud.html.md]: whitespaces

File: website/source/docs/builders/hetzner-cloud.html.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+-   `location` (string) - The name of the location to launch the
+    droplet in.
+
+-   `server_type` (string) - ID or name of the server type this server should be created with.
+
+### Optional:
+
+-   `endpoint` (string) - Non standard api endpoint URL. Set this if you are
+    using a DigitalOcean API compatible service. It can also be specified via
```
Comment (rickard-von-essen) [inline website/source/docs/builders/hetzner-cloud.html.md]: DI

File: website/source/docs/builders/hetzner-cloud.html.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    environment variable `HCLOUD_ENDPOINT`.
+
+-   `server_name` (string) - The name assigned to the droplet. DigitalOcean
+    sets the hostname of the machine to this value.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/hetzner-cloud.html.md]: DI and droplet?

File: website/source/docs/builders/hetzner-cloud.html.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+-   `snapshot_name` (string) - The name of the resulting snapshot that will
+    appear in your account. Defaults to "packer-{{timestamp}}" (see
+    [configuration templates](/docs/templates/engine.html) for more info).
+
+-   `poll_interval` (string) - Configures the interval in which actions are polled by the client. Default `500ms`. Increase this interval if you run into rate limiting errors.
+
+-   `user_data` (string) - User data to launch with the Droplet.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/hetzner-cloud.html.md]: Droplet

File: website/source/docs/builders/hetzner-cloud.html.md (hunk 1 above)
Comment (LKaemmerling) [reply website/source/docs/builders/hetzner-cloud.html.md]: Fixed, thank you!

File: website/source/docs/builders/hetzner-cloud.html.md (hunk 2 above)
Comment (LKaemmerling) [reply website/source/docs/builders/hetzner-cloud.html.md]: Fixed, thank you!

File: website/source/docs/builders/hetzner-cloud.html.md (hunk 3 above)
Comment (LKaemmerling) [reply website/source/docs/builders/hetzner-cloud.html.md]: Fixed, thank you!

File: website/source/docs/builders/hetzner-cloud.html.md (hunk 4 above)
Comment (LKaemmerling) [reply website/source/docs/builders/hetzner-cloud.html.md]: Fixed, thank you!

File: website/source/docs/builders/hetzner-cloud.html.md (hunk 5 above)
Comment (LKaemmerling) [reply website/source/docs/builders/hetzner-cloud.html.md]: Fixed, thank you!

File: website/source/docs/builders/hetzner-cloud.html.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+[communicator](/docs/templates/communicator.html) can be configured for this
+builder.
+
+### Required:
+
+-   `token` (string) - The client TOKEN to use to access your account. It
+    can also be specified via environment variable `HCLOUD_TOKEN`,
+    if set.
+
+-   `image` (string) - ID or name of image to create server in.
```
Comment (SwampDragons) [inline website/source/docs/builders/hetzner-cloud.html.md]: This line doesn't make a ton of sense to me. Maybe "ID or name of image to launch server from." ?

File: builder/hcloud/step_create_server.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+		contents, err := ioutil.ReadFile(c.UserDataFile)
+		if err != nil {
+			state.Put("error", fmt.Errorf("Problem reading user data file: %s", err))
+			return multistep.ActionHalt
+		}
+
+		userData = string(contents)
+	}
+
+	serverCreateResult, _, err := client.Server.Create(context.TODO(), hcloud.ServerCreateOpts{
```
[the remark this answers is not in the document]
Comment (LKaemmerling) [reply builder/hcloud/step_create_server.go]: I think i would leave it as is.

File: builder/hcloud/step_create_server.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		return multistep.ActionHalt
+	}
+	state.Put("server_ip", serverCreateResult.Server.PublicNet.IPv4.IP.String())
+	// We use this in cleanup
+	s.serverId = serverCreateResult.Server.ID
+
+	// Store the server id for later
+	state.Put("server_id", serverCreateResult.Server.ID)
+
+	_, errCh := client.Action.WatchProgress(context.TODO(), serverCreateResult.Action)
```
[the remark this answers is not in the document]
Comment (LKaemmerling) [reply builder/hcloud/step_create_server.go]: I think i would leave it as it is :)

**Comment by SwampDragons** [comment]
This is great; thanks!

**Pull request merged**